packages/headless/src/app/search-engine/jwt-reducer.ts

```diff
@@ -0,0 +1,219 @@
+  );
+  logger.warn(
+    `To remove this warning, make sure that access token value [${tokenValue}] matches engine configuration value [${stateProp}]`
+  );
+};
+
+const shouldReconcileValues = (
+  tokenValue: string | undefined,
+  stateValue: string
+): boolean => {
+  if (isNullOrUndefined(tokenValue)) {
+    return false;
+  }
+
+  if (stateValue === tokenValue) {
+    return false;
+  }
+
+  return true;
+};
+
+const decodeJSONWebToken = (token: string): CoveoJSONWebToken | false => {
+  try {
+    const atobImplementation = typeof atob !== 'undefined' ? atob : atobshim;
+    const base64Url = token.split('.')[1];
+    const base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
+    const base64decoded = atobImplementation(base64);
+    if (!base64decoded) {
+      return false;
+    }
+    const jsonPayload = decodeURIComponent(
+      base64decoded
+        .split('')
+        .map((character) => {
+          return '%' + ('00' + character.charCodeAt(0).toString(16)).slice(-2);
+        })
+        .join('')
+    );
+
+    return JSON.parse(jsonPayload) as CoveoJSONWebToken;
+  } catch (e) {
+    return false;
+  }
+};
+
+const handleMismatchOnSearchHub = (
+  jwt: CoveoJSONWebToken,
+  state: SearchAppState,
+  payload: string | undefined,
+  logger: P.Logger
+): SearchAppState => {
+  possiblyWarnOnMismatch(
+    jwt,
+    'searchHub',
+    state.searchHub,
+    getSearchHubInitialState(),
+    payload,
+    logger
+  );
+  if (shouldReconcileValues(jwt.searchHub, state.searchHub)) {
+    return {
+      ...state,
+      searchHub: jwt.searchHub!,
+    };
+  }
+
+  return state;
+};
+
+const handleMismatchOnPipeline = (
+  jwt: CoveoJSONWebToken,
+  state: SearchAppState,
+  payload: string | undefined,
+  logger: P.Logger
+): SearchAppState => {
+  possiblyWarnOnMismatch(
+    jwt,
+    'pipeline',
+    state.pipeline,
+    getPipelineInitialState(),
+    payload,
+    logger
+  );
+  if (shouldReconcileValues(jwt.pipeline, state.pipeline)) {
+    return {
+      ...state,
+      pipeline: jwt.pipeline!,
+    };
+  }
+  return state;
+};
+
+const handleMismatchOnUserDisplayName = (
+  jwt: CoveoJSONWebToken,
+  state: SearchAppState,
+  payload: string | undefined,
+  logger: P.Logger
+): SearchAppState => {
+  possiblyWarnOnMismatch(
+    jwt,
+    'userDisplayName',
+    state.configuration.analytics.userDisplayName,
+    getConfigurationInitialState().analytics.userDisplayName,
+    payload,
+    logger
+  );
+  if (
+    shouldReconcileValues(
+      jwt.userDisplayName,
+      state.configuration.analytics.userDisplayName
+    )
+  ) {
+    return {
+      ...state,
+      configuration: {
+        ...state.configuration,
+        analytics: {
+          ...state.configuration.analytics,
+          userDisplayName: jwt.userDisplayName!,
+        },
+      },
+    };
+  }
+  return state;
+};
+
+export const jwtReducer: (logger: P.Logger) => Reducer = (logger) => {
```

Comment:
Why not using the `createReducer` toolkit method like for other reducers?

Reply:
The main one was that we don't really have "the initial search app state" (that is needed by createReducer), just the initial state for every individual slice.

I think it would still work if I just use an empty object and cast to SearchAppState however. Or I could create a initial global search app state function.

I'll try that approach.